### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by hkup859**
Changes Made:

1. Run the code 10k times or up to 1 minute, throw out the first 100 results, and average the remaining. Times are significantly reduced due to caching and other effects or processing the same data multiple times, but they are much more consistent. 

2. Added an execution versioning system to track what method a given executionTime was generated with, the new version being 1.1.

3. Added verbage to explain the performance analysis process and the different versions (currently only v1.0 & v1.1).

4. Added code to prevent accidental double or multi-submit due to the now longer runtimes. The button changes to yellow and says "Processing...". You cannot submit until either it finishes or unless you refresh the page. A queue system will be better, but this is a simple stopgap until one is implemented.

Feel free to recommend additional changes. I took a few liberties, but am hoping this is a reasonable update that we can build off of. I'm not the greatest with UI, so if there's a particular way you want something to look let me know.

### CODE REVIEW

**Comment by sigmaSd** [comment]
This preview build is nice I can test without pulling the changes locally
image

File: main.ts [hunk 1]
```diff
@@ -225,12 +226,28 @@ async function handleRequest(req: Request): Promise<Response> {
...
+        const allExecutionTimes: any[] = [];
+        const executionStartTimestamp = Date.now();
+        const FIVE_MINUTES = 1 * 60 * 1000;
+        for (let i = 0; i < 10000; i++) {
+          const runResult = await runCode(data.code, day, part);
+          // Discard the first 100 runs as a "warm up"
+          if (i > 100) allExecutionTimes.push(runResult);
+
+          // Exit if runs are taking more than 5 minutes
+          if (Date.now() - executionStartTimestamp > FIVE_MINUTES) i = Infinity;
```
Comment (sigmaSd) [inline main.ts:238]: Why not `break` instead, I saw you also did this in your aoc code

**Comment by sigmaSd** [comment]
The commits are clear thanks, I think the  processing button should be more descriptive though like
"The server is currently processing your request (this may take up to 5 minutes), please wait..."

**Comment by sigmaSd** [comment]
Sometimes I get this error
image

**Comment by sigmaSd** [comment]
Lets go with this and improve from there, I left a couple of comments on this pr

**Comment by sigmaSd** [comment]
@hkup859 I see the problem now we're surpassing deploy cpu limits, 
image


i decreased the iteration number to 1000

**Comment by sigmaSd** [comment]
Yeah thats not enough, its unfortante, but for now i just made it run the code 10 times only, v1.2

**Comment by sigmaSd** [comment]
I'll see if I can increase more later, but I'm starting with the safest lowest minimum, I think if we implement the feature of only running one code at a time, we can increase safely until we reach the maximum possible

**Comment by hkup859** [comment]
Interesting, this is deployed to deno.dev free tier correct? I will try to reproduce the CPU issue and see if there is a way around it. 

I also see I left the code saying FIVE_MiINUTES, but the actual check is only for 1 minute (which was intentional). I'll fix that in my next pr.

**Comment by hkup859** [comment]
At first glance, it looks like we can only have 50ms CPU time per request. So to get any useful amount of runs we will need to implement queueing and execute runs as separate requests. 

Deno has an email to contact if your use case exceeds their limits. I'm gonna try that, but I expect we'll need queueing.

UPDATE: Based on some rough math, I don't think we'll be able to do more processing based on queueing either. Queueing would be dependent on writes to KV and at 300k writes per month, that would only allow 100 runs for 100 people/code solutions across 30 days. I suppose at the moment the site isn't getting that much traffic, but it doesn't allow for much flexibility. We'll see if Deno gets back to me on increasing the CPU time, otherwise I may try some deploy testing with AWS as the free tier ec2 should work well for this use case.

UPDATE 2: I did some testing on my own copy of this repo deployed to deno.dev and I have no idea how they are calculating CPU time. I got tests up to 7 seconds to run without throwing an error. More curiously, I had requests running for about 10-13 seconds that would succeed, save, return, then a couple seconds later fail in the logs. I wonder if they have some bugs related to their CPU monitoring. Anyways, will have to look at this more another time. For now I think the best course of action might be to offload the code running to a reliable, or at least more flexible, service. Currently thinking of AWS ec2 or lambda. The main repo could still be hosted in deno.dev and we'd keep the KV data, just the processing would get moved to AWS.

**Pull request merged**